File: cap10/src/algoritmos/BusquedaBoyerMoore.java
package algoritmos;

import java.util.ArrayList;
import java.util.List;

public class BusquedaBoyerMoore {
	private static final int NUM_CHARS = 256;

	public static int[] tablaUltimaPosicion(String patron) {
		int[] ultPos;
		ultPos = new int[NUM_CHARS];
		for (int c = 0; c < NUM_CHARS; c++) {
			ultPos[c] = -1;
		}
		for (int j = 0; j < patron.length(); j++) {
			ultPos[patron.charAt(j)] = j;
		}
		return ultPos;
	}

	public static List<Integer> buscar(String texto, String buscado) {
		List<Integer> sol = new ArrayList<>();
		int longTexto = texto.length();
		int longBusc = buscado.length();
		int[] ultPos = tablaUltimaPosicion(buscado);

		int i = 0;
		while (i <= longTexto - longBusc) {
			int j = longBusc - 1;
			while (j >= 0 && buscado.charAt(j) == texto.charAt(i + j)) {
				j--;
			}
			if (j < 0) {
				sol.add(i);
				if (i + longBusc < longTexto) {
					i += longBusc - ultPos[texto.charAt(i + longBusc)];
				} else {
					i++;
				}
			} else {
				if (j - ultPos[texto.charAt(i + j)] > 1) {
					i += j - ultPos[texto.charAt(i + j)];
				} else {
					i++;
				}
			}
		}
		return sol;
	}

	public static void main(String[] args) {
		System.out.println(buscar("abracadabra", "bra"));
	}
}

File: cap10/src/algoritmos/BusquedaBoyerMoore.py
NUM_CHARS = 256
 
def tablaUltimaPosicion(patron):
    ultPos = [-1] * NUM_CHARS
    for i in range(len(patron)):
        ultPos[ord(patron[i])] = i
    return ultPos
 
def buscar(texto, buscado):
    sol = []
    longBusc = len(buscado)
    longTexto = len(texto)

    ultPos = tablaUltimaPosicion(buscado)
    i = 0

    while i <= longTexto - longBusc:
        j = longBusc - 1
        while j >= 0 and buscado[j] == texto[i+j]:
            j -= 1 # mientras coincidan, vamos retrocediendo
        
        if j < 0: # si coinciden todas, lo hemos encontrado
            sol.append(i)
            if i + longBusc < longTexto:
                i += longBusc - ultPos[ord(texto[i+longBusc])] 
            else:
                i += 1
        else: # no lo hemos encontrado, saltamos
            i += max(1, j - ultPos[ord(texto[i+j])])
    return sol
 
print(buscar("abracadabra", "bra"))